homeassistant/helpers/storage.py

```diff
@@ -133,6 +134,10 @@ async def _async_load_data(self):
             # If we didn't generate data yet, do it now.
             if "data_func" in data:
                 data["data"] = data.pop("data_func")()
+
+            # We make a copy because code might assume it's safe to mutate loaded data
+            # and we don't want that to mess with what we're trying to store.
+            data = deepcopy(data)
```

Comment:
`deepcopy` could be expensive here if the data is large.

Is it possible to use `MappingProxyType` instead?

Reply:
Not sure if this case actually happens enough that it matters though

Reply:
Doing some testing now....

Reply:
This doesn't seem to be a problem in practice.